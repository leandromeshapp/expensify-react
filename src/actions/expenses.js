import uuid from "uuid"
import database from "../firebase/firebase"



// ADD EXPENSE
export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense,
})


export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        const {
            description = "", 
            note = "", 
            amount = 0, 
            createdAt = 0 
        } = expenseData
        const expense = { description, note, amount, createdAt }
        return database.ref(`users/${uid}/expenses`).push(expense)
        .then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense,
            }))
        })
    }
}


// REMOVE EXPENSE
    export const removeExpense = id => ({
    type: "REMOVE_EXPENSE",
    id,
})


export const startRemoveExpense = id => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses/${id}`)
        .remove().then(() => {
            dispatch(removeExpense(id))
        })
    }
}




// EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
})


export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses/${id}`)
        .update(updates)
        .then(() => {
            dispatch(editExpense(id, updates));
        });
    };
  };


// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
})


export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses`).once("value")
        .then((snapshot) => {
            const expenses = []

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}



// SET_order_expenses
export const setOrderExpenses = (expenses) => ({
    type: "SET_ORDER_EXPENSES",
    expenses
})


export const orderExpenses = (expenses) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth
        return database.ref(`users/${uid}/expenses`)
        .orderByKey()
        .on("child_added", callback)
        .limitToFirst(5)
        .then(() => {
            dispatch(setOrderExpenses(expenses))
        })
    }
}