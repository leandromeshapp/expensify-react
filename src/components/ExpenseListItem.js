import React from "react"
import { Link } from "react-router-dom"
import moment from "moment"
import numeral from "numeral"
//import locale from 'numeral/locales';

//numeral.locale('pt-pt');
const ExpenseListItem = ({ id, description, amount, createdAt, note }) => (
    <Link className="list-item" to={`/edit/${id}`}>
        <div>
            <h3 className="list-item__title"> 
                { description } 
            </h3>

            <span className="list-item__sub-title"> { moment(createdAt).format("D MMMM YYYY") } </span>
        </div>
        <span className="list-item__note">{note}</span>
        <h3 className="list-item__data"> { numeral(amount / 100).format("0,0.00$") }  </h3>
    </Link>
)

export default ExpenseListItem