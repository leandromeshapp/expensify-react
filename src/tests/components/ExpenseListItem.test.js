import React from "react"
import { shallow } from "enzyme"
import expenses from "../fixtures/expenses"
import ExpenseListItem from "../../components/ExpenseListItem"



// Render ExpenseListItem Component with fixture data 
test("Should render ExpenseListItem with fixture data", () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
    expect(wrapper).toMatchSnapshot()
})