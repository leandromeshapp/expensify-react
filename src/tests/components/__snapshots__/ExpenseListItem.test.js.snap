// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListItem with fixture data 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Gum
    </h3>
    <span
      className="list-item__sub-title"
    >
       
      1 January 1970
       
    </span>
  </div>
  <h3
    className="list-item__data"
  >
     
    $1.00
      
  </h3>
</Link>
`;
